File: Frontend/static/js/server_configuration.js
$(document).ready(function () {

    checkServerStatus();

    loadUploadedFilesAndEndpoint();

    fetchFacilityList()
    
    // Call update function for each file input
    updateFileInfo("file1", "fileInfo1");
    updateFileInfo("file2", "fileInfo2");
    updateFileInfo("file3", "fileInfo3");

    loadExistingFiles();

    setInterval(checkServerSession, 5000);

    // Start Server Button Click
    $("#startServer").click(function () {
        $.post("/start_server", function (response) {
            $("#serverMessage").html("<p style='color: green;'>" + response.message + "</p>");
            checkServerStatus();
        }).fail(function () {
            $("#serverMessage").html("<p style='color: red;'>Error starting the server.</p>");
        });
    });

    // Stop Server Button Click
    $("#stopServer").click(function () {
        $.get("/stop_server", function (response) {
            $("#serverMessage").html("<p style='color: green;'>" + response.message + "</p>");
            checkServerStatus();
        }).fail(function () {
            $("#serverMessage").html("<p style='color: red;'>Error stopping the server.</p>");
        });
    });

    // Handle Authentication Form Submit
    $("#AuthenticateForm").submit(function (event) {
        event.preventDefault(); // Prevent page reload

        $("#AuthenticateButton").prop("disabled", true).text("Authenticating..."); // Disable button

        $.post("/authenticate", $(this).serialize(), function (response) {
            if (response.success) {
                // Login successful → Redirect
                $("#authMessage").css("color", "green").text("Login successful! Redirecting...");
                setTimeout(() => {
                    window.location.href = response.redirect;
                }, 1000); // Redirect after 1 second
            } else {
                // Wrong login → Show error message instead of redirecting
                $("#authMessage").css("color", "red").text("Wrong username or password.");
                $("#AuthenticateButton").prop("disabled", false).text("Authenticate"); // Re-enable button
            }
        }).fail(function (xhr) {
            // Server error → Show error instead of redirecting
            $("#authMessage").css("color", "red").text("Server error! Please try again.");
            $("#AuthenticateButton").prop("disabled", false).text("Authenticate");
        });
    });

    // Handle Facility ID Submission
    $("#facilityForm").submit(function (event) {
        event.preventDefault(); // Prevent page reload

        var selectedId = $("#facility_id").val(); // Get selected facility ID

        if (!selectedId) {
            $("#facilityMessage").html("<p style='color: red;'>Please select a Facility ID!</p>");
            return;
        }

        $.post("/select_facility_id", { facility_id: selectedId }, function (response) {
            if (response.success) {
                $("#facilityMessage").html("<p style='color: green;'>Facility ID = \"" + selectedId + "\" updated successfully!</p>");
            } else {
                $("#facilityMessage").html("<p style='color: red;'>" + response.message + "</p>");
            }
        }).fail(function () {
            $("#facilityMessage").html("<p style='color: red;'>Error updating Facility ID.</p>");
        });
    });

    // Update displayed file name when user selects a file
    $("#file1").change(function () {
        $("#fileInfo1").text(this.files.length ? this.files[0].name : "No file chosen");
    });
    $("#file2").change(function () {
        $("#fileInfo2").text(this.files.length ? this.files[0].name : "No file chosen");
    });
    $("#file3").change(function () {
        $("#fileInfo3").text(this.files.length ? this.files[0].name : "No file chosen");
    });
    
    // Handle Certificate Upload
    $("#uploadForm").submit(function (event) {
        event.preventDefault(); // Prevent page refresh
        var formData = new FormData(this);

        $.ajax({
            url: "/upload_cert",
            type: "POST",
            data: formData,
            processData: false,
            contentType: false,
            success: function (response) {
                if (response.success) {
                    // Update file paths dynamically based on uploaded files
                    if (response.updated_paths.file1) {
                        $("#fileInfo1").text(response.updated_paths.file1);
                    }
                    if (response.updated_paths.file2) {
                        $("#fileInfo2").text(response.updated_paths.file2);
                    }
                    if (response.updated_paths.file3) {
                        $("#fileInfo3").text(response.updated_paths.file3);
                    }

                    // Display success message
                    $("#uploadMessage").html("<p style='color: green;'>Uploaded: " + response.uploaded_files.map(f => `"${f}"`).join(", ") + " successfully!</p>");

                    // Refresh the uploaded files list dynamically
                    loadUploadedFilesAndEndpoint();

                } else {
                    // Display error message for invalid file type
                    $("#uploadMessage").html("<p style='color: red;'>" + response.message + "</p>");
                }

                // Clear file input fields
                $("#file1").val(""); 
                $("#file2").val(""); 
                $("#file3").val(""); 
            },
            error: function (xhr) {
                // Display error message for invalid file type
                var errorMessage = xhr.responseJSON ? xhr.responseJSON.message : "Error uploading file.";
                $("#uploadMessage").html("<p style='color: red;'>" + errorMessage + "</p>");
            }
        });
    });

    // Handle AWS IoT Endpoint Update
    $("#awsEndpointForm").submit(function (event) {
        event.preventDefault();
        let newEndpoint = $("#aws_iot_endpoint").val().trim();

        if (!newEndpoint) {
            $("#AWSEndPoint").html("<p style='color: red;'>AWS IoT Endpoint cannot be empty!</p>");
            return;
        }

        $.post("/update_aws_endpoint", { aws_iot_endpoint: newEndpoint })
            .done(function (response) {
                $("#AWSEndPoint").html(`<p style='color: green;'>AWS IoT Endpoint updated successfully!</p>`);
                $("#awsEndpointDisplay").text(response.new_endpoint);
                $("#aws_iot_endpoint").val("");
                
                loadUploadedFilesAndEndpoint()
            })
            .fail(function () {
                $("#AWSEndPoint").html("<p style='color: red;'>Error updating AWS IoT Endpoint.</p>");
            });
    });

    // Handle AWS Connection Check
    $("#AWSCheckingButton").click(function () {
        console.log("Checking AWS connection...");

        $("#AWSEndPoint").html("<p style='color: blue;'>Checking connection...</p>");
        $("#AWSCheckingButton").prop("disabled", true);

        $.get("/check_aws_connection")
            .done(function (response) {
                $("#AWSCheckingButton").prop("disabled", false);
                let statusText = {
                    success: "<p style='color: green;'>Connection successful!</p>",
                    failure: "<p style='color: red;'>Connection failed!</p>",
                    timeout: "<p style='color: orange;'>Connection check timed out!</p>"
                };
                $("#AWSEndPoint").html(statusText[response.status] || "<p style='color: red;'>Unknown status.</p>");
            })
            .fail(function () {
                $("#AWSEndPoint").html("<p style='color: red;'>Request error calling /check_aws_connection</p>");
            });
    });

    // Handle AWS IoT Connection Check Button Click
    $("#AWSCheckingButton").click(function () {
        console.log("Check Connection button clicked!"); // Debugging

        // 1) Clear old status text
        $("#AWSEndPoint").empty();
        
        // 2) Show a "checking..." message or spinner
        $("#loadingSpinner").show();

        // Disable the button so it can’t be clicked again during the check
        $(this).prop("disabled", true);

        // Send request to check AWS connection
        $.get("/check_aws_connection", function (response) {
            // Hide spinner once we have a response
            $("#loadingSpinner").hide();
            $("#AWSCheckingButton").prop("disabled", false); // Re-enable button

            // Handle response status
            let message = "";
            if (response.status === "success") {
                message = "<p style='color: green;'>Connection successful!</p>";
            } else if (response.status === "failure") {
                message = "<p style='color: red;'>Connection failed!</p>";
            } else if (response.status === "timeout") {
                message = "<p style='color: orange;'>Connection check timed out!</p>";
            } else {
                message = "<p style='color: red;'>Unknown status returned.</p>";
            }

            // Update the status message
            $("#AWSEndPoint").html(message);
        })
        .fail(function () {
            $("#loadingSpinner").hide();
            $("#AWSEndPoint").html("<p style='color: red;'>Request error calling /check_aws_connection</p>");
        });
    });
})

// Fetch Facility List from API and Update UI
function fetchFacilityList() {
    console.log("Fetching facility list..."); // Debugging

    $.get("/facility_list", function(response) {
        console.log("Facility List Response:", response); // Debugging

        if (response.success) {
            var tableBody = $("#facilityTable tbody");
            tableBody.empty(); // Clear existing rows

            response.data.forEach(function(item) {
                var row = `
                    <tr>
                        <td>${item.id}</td>
                        <td>${item.name}</td>
                        <td>${item.address}</td>
                        <td>${item.timezone}</td>
                        <td>${item.tcp_server_name}</td>
                    </tr>
                `;
                tableBody.append(row);
            });

            $("#facilityCount").text(response.count);
        } else {
            $("#facilityMessage").html("<p style='color: red;'>No facilities found.</p>");
        }
    }).fail(function() {
        $("#facilityMessage").html("<p style='color: red;'>Error fetching facility list.</p>");
    });
}

// Update displayed file name when a new file is selected
function updateFileInfo(inputId, displayId) {
    $("#" + inputId).change(function () {
        if (this.files.length > 0) {
            let fileName = this.files[0].name;
            $("#" + displayId).text(fileName); // Show just file name before upload
        }
    });
}

// Load existing file paths from the server
function loadExistingFiles() {
    $.get("/get_existing_files", function (response) {
        if (response.success) {
            $("#fileInfo1").text(response.root_ca_path);
            $("#fileInfo2").text(response.private_key_path);
            $("#fileInfo3").text(response.cert_path);
        } else {
            console.error("Failed to load existing files.");
        }
    }).fail(function () {
        console.error("Error fetching existing files.");
    });
}

// Function to refresh the uploaded files and AWS IoT Endpoint display dynamically
function loadUploadedFilesAndEndpoint() {
    $.get("/get_uploaded_files", function (response) {
        var fileListElement = $("#uploadedFilesList");
        fileListElement.empty();

        if (response.files.length > 0) {
            response.files.forEach(function (file) {
                fileListElement.append("<li>" + file + "</li>");
            });
        } else {
            fileListElement.append("<li>No files uploaded</li>");
        }

        // Update AWS IoT Endpoint in "Uploaded Files & AWS IoT Endpoint" section
        $("#awsEndpointDisplay").text(response.aws_endpoint);
    }).fail(function () {
        console.error("Error fetching uploaded files and endpoint.");
    });
}

function checkServerStatus() {
    $.get("/button_server_status", function (response) {
        console.log("Server status response:", response); // Debugging

        if (response.running) {
            // Server is running → Disable everything except Stop Server
            $("#startServer").prop("disabled", true).css("opacity", "0.5");
            $("#stopServer").prop("disabled", false).css("opacity", "1");

            // Disable all other buttons
            $("#AuthenticateButton").prop("disabled", true).css("opacity", "0.5"); 
            $("#AWSUpdateButton").prop("disabled", true).css("opacity", "0.5");
            $("#AWSCheckingButton").prop("disabled", true).css("opacity", "0.5");
            $("#UploadButton").prop("disabled", true).css("opacity", "0.5");
            $("#FacilityIDButton").prop("disabled", true).css("opacity", "0.5"); // Disable Select Facility Submit Button

        } else {
            // Server is stopped → Enable everything
            $("#stopServer").prop("disabled", true).css("opacity", "0.5");
            $("#AuthenticateButton").prop("disabled", false).css("opacity", "1"); 
            $("#AWSUpdateButton").prop("disabled", false).css("opacity", "1");
            $("#AWSCheckingButton").prop("disabled", false).css("opacity", "1");
            $("#UploadButton").prop("disabled", false).css("opacity", "1");
            $("#FacilityIDButton").prop("disabled", false).css("opacity", "1"); // Enable Select Facility Submit Button

            // Enable Start Server only if allowed
            if (response.enable_start_button) {
                $("#startServer").prop("disabled", false).css("opacity", "1"); // Enable Start Server if allowed
            } else {
                $("#startServer").prop("disabled", true).css("opacity", "0.5"); // Keep Start Server disabled
            }
        }
    }).fail(function () {
        console.error("Failed to fetch server status");
    });
}

function checkServerSession() {
    $.get("/check_session", function(response) {
        if (!response.valid) {
            alert(response.message);

            // Disable buttons and prevent further actions
            $("button").prop("disabled", true).css("opacity", "0.5");

            // Redirect user to login page after 1 second
            setTimeout(function() {
                window.location.href = "/login";
            }, 1000);
        }
    }).fail(function () {
        console.error("Error checking session.");
    });
}

File: Backend/app/routes/api/server_configuration.py
from flask import Blueprint, session, redirect, url_for, render_template, jsonify, request

import app.services.control_server as control_server
import app.environment.environment_manager as environment_manager
import app.services.auth_server as auth_server
import app.routes.api.system_log as system_log

import os
import asyncio

server_config_bp = Blueprint("server_configuration", __name__)

UPLOAD_FOLDER = "Backend/app/upload"

@server_config_bp.route("/server_configuration")
def server_configuration():
    if not session.get("authenticated"):
        return render_template("login.html")

    if session.get("username") == "supervisor":

        print("[WARNING] Unauthorized access attempt by Supervisor.")
        system_log.log_to_redis("[WARNING] Unauthorized access attempt by Supervisor.")
        
        return redirect(url_for("server_management.server_management"))

    if session.get("username") == "admin" and not auth_server.validate_admin_session():

        print("[WARNING] Admin session invalid. Forcing logout.")
        system_log.log_to_redis("[WARNING] Admin session invalid. Forcing logout.")

        session.clear()
        return render_template("login.html")

    # return render_template("server_configuration.html")
    facilities = environment_manager.get_facility_list()

    if not facilities or "data" not in facilities:
        facilities = {"data": [], "count": 0}  # Prevent errors

    return render_template("server_configuration.html", data=facilities["data"], count=facilities["count"])

# ------------------------------------------------------
# Start TCP server and accept data from devices 
# ------------------------------------------------------ 
@server_config_bp.route("/start_server", methods=["GET", "POST"])
def start_server():

    environment_manager.update_start_server_status(True)
    environment_manager.update_stop_server_status(False)

    return jsonify({"success": True, "message": "TCP Server started successfully!"})
    
# ------------------------------------------------------
# Stop TCP server and not accept data from devices 
# Stop TCP server to config server
# ------------------------------------------------------     
@server_config_bp.route("/stop_server", methods=["GET"])
def stop_server():

    environment_manager.update_start_server_status(False)
    environment_manager.update_stop_server_status(True)

    return jsonify({"success": True, "message": "TCP Server stopped successfully!"})

# --------------------------------------------------
# Authenticate Function
# --------------------------------------------------
@server_config_bp.route("/authenticate", methods=["POST"])
def authenticate():

    # Extract username & password
    username = request.form.get("username", "").strip()
    password = request.form.get("password", "").strip()

    print(f"[INFO] Received authenticate request - Username: {username}, Password: {'*' * len(password)}")
    system_log.log_to_redis(f"[INFO] Received authenticate request - Username: {username}, Password: {'*' * len(password)}")
    # Call authentication service
    success, redirect_url = auth_server.authenticate_user(username, password)

    return jsonify({"success": success, "redirect": redirect_url})

# --------------------------------------------------
# Facility list Function
# --------------------------------------------------
@server_config_bp.route("/facility_list", methods=["GET"])
def facility_list():

    facilities = auth_server.facility_list()

    if not facilities:
        return jsonify({"success": False, "message": "Authentication failed or no facilities available"}), 403

    return jsonify({"success": True, "data": facilities["data"], "count": facilities["count"]})

# ------------------------------------------------------
# Select Facility ID - Update Environment & Confirm Selection
# ------------------------------------------------------
@server_config_bp.route("/select_facility_id", methods=["POST"])
def select_facility_id():

    try:
        facility_id = request.form.get("facility_id")

        if not facility_id:
            return jsonify({"success": False, "message": "Facility ID not provided!"}), 400

        # Get the facility list
        facilities = environment_manager.get_facility_list()

        # Check if facility exists
        facility_exists = any(str(fac["id"]) == facility_id for fac in facilities.get("data", []))

        if not facility_exists:
            return jsonify({"success": False, "message": "Facility ID does not exist!"}), 404

        # Update the selected Facility ID in environment
        environment_manager.update_facility_id(facility_id)

        print(f"[INFO] Facility ID Updated: {facility_id}")
        system_log.log_to_redis(f"[INFO] Facility ID Updated: {facility_id}")

        return jsonify({"success": True, "message": f"Facility ID '{facility_id}' updated successfully!"})

    except Exception as e:
        return jsonify({"success": False, "message": f"Error: {str(e)}"}), 500

# --------------------------------------------------
# Upload Certificates function: 
# -> Get 3 files of certificates and stored it in 
# upload folder
# --------------------------------------------------
@server_config_bp.route("/upload_cert", methods=["POST"])
def upload_cert():
    file_mapping = {
        "file1": "root_ca",
        "file2": "private_key",
        "file3": "cert"
    }

    saved_files = []
    updated_paths = {}

    for key, file_type in file_mapping.items():
        file = request.files.get(key)

        if not file or file.filename == '':
            continue

        ALLOWED_EXTENSIONS = environment_manager.get_allow_extensions()

        file_ext = os.path.splitext(file.filename)[1]  # Extract the extension correctly

        # Check if the extension is allowed
        if not any(file.filename.endswith(ext) for ext in ALLOWED_EXTENSIONS):
            return jsonify({
                "success": False,
                "message": f"Invalid file type: {file.filename}. Allowed types: {', '.join(ALLOWED_EXTENSIONS)}"
            }), 400

        try:
            # Delete all existing files with the same extension
            for existing_file in os.listdir(UPLOAD_FOLDER):
                if existing_file.endswith(file_ext):  # Match only by file extension
                    os.remove(os.path.join(UPLOAD_FOLDER, existing_file))
                    print(f"Deleted old file: {existing_file}")  # Debugging
                    system_log.log_to_redis(f"Deleted old file: {existing_file}")

            file_path = os.path.join(UPLOAD_FOLDER, file.filename)
            file.save(file_path)

            updated_paths[key] = file_path.replace("\\", "/")

            updated_paths[key] = environment_manager.update_environment_file(file_path, file_type)

            saved_files.append(file.filename)

            environment_manager.update_status_aws_server(False)

        except Exception as e:
            return jsonify({"success": False, "error": str(e)}), 500

    if saved_files:
        return jsonify({"success": True, "uploaded_files": saved_files, "updated_paths": updated_paths}), 200
    else:
        return jsonify({"success": False, "message": "No files selected"}), 400

# ---------------------------------------------------
# Update AWS IoT Endpoint function: 
# -> Get endpoint from GUI and update to 
# Environment.py
# ---------------------------------------------------
@server_config_bp.route("/update_aws_endpoint", methods=["POST"])
def update_aws_endpoint_route():

    new_endpoint = request.form.get("aws_iot_endpoint", "").strip()

    if not new_endpoint:
        return jsonify({"success": False, "message": "AWS IoT Endpoint cannot be empty!"}), 400

    environment_manager.update_aws_endpoint(new_endpoint)

    return jsonify({"success": True, "new_endpoint": new_endpoint})

# ------------------------------------------------------
# Get existing files and path of files: aws certificates
# ------------------------------------------------------
@server_config_bp.route('/get_existing_files', methods=['GET'])
def get_existing_files():

    # Reload environment.py to get the latest updates
    environment_manager.reload_environment()

    return jsonify({
        "success": True,
        "root_ca_path": environment_manager.get_root_ca_path(),
        "private_key_path": environment_manager.get_private_key_path(),
        "cert_path": environment_manager.get_cert_path()
    })

# ------------------------------------------------------
# Get uploaded files
# ------------------------------------------------------
@server_config_bp.route("/get_uploaded_files", methods=["GET"])
def uploaded_files():

    files, aws_iot_endpoint = environment_manager.get_uploaded_files()  # Get files + endpoint dynamically
    return jsonify({"files": files, "aws_endpoint": aws_iot_endpoint})

# ------------------------------------------------------
# Check AWS Connection function: 
# -> Check connection with new AWS IoT Endpoint
# ------------------------------------------------------
@server_config_bp.route("/check_aws_connection", methods=["GET"])
def check_aws_connection():

    # Run the async function inside an event loop
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    status = loop.run_until_complete(auth_server.process_aws_connection_check())

    return jsonify({"status": status})

# ------------------------------------------------------
# Check the current status of server to disable buttons 
# ------------------------------------------------------
@server_config_bp.route("/button_server_status")
def button_server_status():
    
    status_server = environment_manager.get_start_server()

    if status_server:
        server_running = True
    else:
        server_running = False

    auth_token = environment_manager.get_auth_token()

    if auth_token:
        enable_start_button = True
    else:
        enable_start_button = False

    return jsonify({"running": server_running, "enable_start_button": enable_start_button})  
